sound/soc/sof/ipc4-pcm.c

```diff
@@ -535,13 +539,27 @@ static int sof_ipc4_pcm_dai_link_fixup(struct snd_soc_pcm_runtime *rtd,
 		return -EINVAL;
 	}
 
+
 	for_each_pcm_streams(dir) {
 		struct snd_soc_dapm_widget *w = snd_soc_dai_get_widget(cpu_dai, dir);
 
 		if (w) {
 			struct snd_sof_widget *swidget = w->dobj.private;
 			struct snd_sof_widget *pipe_widget = swidget->spipe->pipe_widget;
 			struct sof_ipc4_pipeline *pipeline = pipe_widget->private;
+			if (dir == SNDRV_PCM_STREAM_PLAYBACK) {
+				if (ipc4_copier->available_fmt.num_output_formats == 1) {
+					ipc4_fmt = &ipc4_copier->available_fmt.
+						   output_pin_fmts->audio_fmt;
+					single_fmt = true;
+				}
+			} else {
+				if (ipc4_copier->available_fmt.num_input_formats == 1) {
+					ipc4_fmt = &ipc4_copier->available_fmt.
+						   input_pin_fmts->audio_fmt;
+					single_fmt = true;
+				}
+			}
```

Comment:
I must admit not following the explanations *at all*.

This code checks for the formats on the dai copier, not the host. So why does the commit message refer to the FE DAI?

In addition, I am struggling to understand the relationship, if any, between the input and output formats of a dai copier, and what would happen if there is more than one format specified. How would that work for the Maxim amplifier which cannot support 24 bits?

Reply:
> This code checks for the formats on the dai copier, not the host. So why does the commit message refer to the FE DAI?

@plbossart Sorry for confusing you. What I am trying to do is to set the BE format as the DAI format that closest to the BE.

Take sof-tgl-rt711-rt1316-rt714.tplg for example,
![sof-tgl-rt711-rt1316-rt714](https://github.com/thesofproject/linux/assets/8856808/8fe7e2a8-1015-4aa3-9756-bf53ef456669)

It is alh-copier.SDW0-Playback.0's output format for Jack out PCM and alh-copier.SDW0-Capture.0's input format for Jack In PCM.

Maybe I should not call `alh-copier.SDW0-Playback.0` and `alh-copier.SDW0-Capture.0` as FE DAI.


> In addition, I am struggling to understand the relationship, if any, between the input and output formats of a dai copier, and what would happen if there is more than one format specified. How would that work for the Maxim amplifier which cannot support 24 bits?

We won't know which format will be used if there are more than one formats available in a copier at this moment. BE format is decided before copier formats are selected. Therefore, we can only fixup to set the BE format if there is only one available format in the DAI copier.
Regarding Maxim amplifier, we will need to either add a new topology with valid_bit_depth = 24, or change valid_bit_depth to 24 in the existing topologies, and make sure the DAI copier supports only one output format.
BTW, that is the reason that I used SOF_TKN_COMP_FORMAT to specify the BE format just like IPC3 does.